slave/slave.go

```diff
@@ -8,17 +8,94 @@ import (
 	"github.com/aaronang/cong-the-ripper/lib"
 )
 
+var slaveInstance Slave
+
+type Slave struct {
+	heartbeat   lib.Heartbeat
+	successChan chan CrackerSuccess
+	failChan    chan CrackerFail
+	addTaskChan chan lib.Task
+}
+
+func Init(instanceId string) *Slave {
+	heartbeat := lib.Heartbeat{
+		SlaveId: instanceId,
+	}
+
+	slaveInstance = Slave{
+		heartbeat:   heartbeat,
+		successChan: make(chan CrackerSuccess),
+		failChan:    make(chan CrackerFail),
+		addTaskChan: make(chan lib.Task),
+	}
+	return &slaveInstance
+}
+
+func (s *Slave) Run() {
+	http.HandleFunc(lib.TasksCreatePath, taskHandler)
+	go http.ListenAndServe(lib.Port, nil)
+
+	for {
+		select {
+		case task := <-s.addTaskChan:
+			s.addTask(task)
+			go s.Execute(task)
+
+		case msg := <-s.successChan:
+			s.password_found(msg.taskID, msg.password)
+
+		case msg := <-s.failChan:
+			s.password_not_found(msg.taskID)
+		}
+	}
+}
+
 func taskHandler(w http.ResponseWriter, r *http.Request) {
```

Comment:
This is a bit of a debatable one. We're using pointer receivers in master, but that's not ideal because the handler runs concurrently with the master loop and it shouldn't have access to all of its fields, only the channels. So what we have here is a actually a bit safer.

To be extra safe, we should change the handlers in master, wrap them in a `makeHandler` function like what is used here  https://golang.org/doc/articles/wiki/final.go